JS/Lesson19/goksel/recursion.js

```diff
@@ -36,103 +36,231 @@ console.log('5', getFactorial(5));
+            return 1 + vowelsInString(sentence.slice(1));
+        } else {
+            return vowelsInString(sentence.slice(1));
+        }
+    }
 }
 console.log(vowelsInString(sentenceEx4));
+
+
+
+console.log("---Ex4---");
 // 4. Write a recursive function to remove all occurrences of a specified character from a string.
 const sentenceEx5 = 'Hello, how are you?';
 const charToRemove = 'o';
 
+function removeCharacter(sentence, removeWord){
+    if(sentence.length === 0 || sentence[0].length === 0){
+        return "";
+        }
+        if(sentence[0] === removeWord ){
+            return removeCharacter(sentence.slice(1), removeWord)
+        }else {
+            return sentence[0] + removeCharacter(sentence.slice(1), removeWord);
+        }
+}
+
+console.log(removeCharacter(sentenceEx5,charToRemove));
+
+console.log("---Ex5---");
 // 5. Write a recursive function to check if an array includes a specific value.
 const numbers = [1, 2, 3, 4, 5];
-const valueToCheck = 3;
+const valueToCheck = 2;
+
+
+function isArrayIncludeValue(array, value){
+    if(array.length === 0){
+        return "Value is not found";
+    } 
+    if(array[0] !== value){
+        return isArrayIncludeValue(array.slice(1), value);
+    } else{
+        return true;
+    }
+
+}
 
+console.log(isArrayIncludeValue(numbers, valueToCheck));
+console.log("---Ex6---");
 // 6. Write a recursive function to flatten an object with nested objects into a single-level object.
 const nestedObj = {
-	name: 'John',
-	age: 30,
-	address: {
-		street: '123 Main St',
-		city: 'New York',
-		country: 'USA',
-	},
-	hobbies: ['reading', 'coding'],
-	social: {
-		twitter: '@johndoe',
-		linkedIn: 'johndoe',
-	},
+    name: 'John',
+    age: 30,
+    address: {
+        street: '123 Main St',
+        city: 'New York',
+        country: 'USA',
+    },
+    hobbies: ['reading', 'coding'],
+    social: {
+        twitter: '@johndoe',
+        linkedIn: 'johndoe',
+    },
 };
 
+
+
+// I have no idea about this example...
+
+console.log("---Ex7---");
 // 7. Write a recursive function to find the maximum depth of a nested object.
 const nestedObjEx7 = {
-	a: 1,
-	b: {
-		c: 2,
-		d: {
-			e: 3,
-			f: {
-				g: 4,
-			},
-		},
-	},
-	h: 5,
+    a: 1,
+    b: {
+        c: 2,
+        d: {
+            e: 3,
+            f: {
+                g: 4,
+            },
+        },
+    },
+    h: 5,
 };
+
+
+function maxDepthOfObject(obj, depth = 0) {
+    if (typeof obj !== "object" || obj === null){
+        return depth;
+    }
+
+    let maxDepth = depth;
+    for(let key in obj){
+        if(obj.hasOwnProperty(key)){
+            let currentDepth = maxDepthOfObject(obj[key], depth +1);
+            if (currentDepth > maxDepth){
+                maxDepth = currentDepth;
+            }
+        }
+    }
+    return maxDepth;
+}
+
+console.log(maxDepthOfObject(nestedObjEx7));
+console.log("---Ex8---");
+
+// I checked this solution from chatgbt, ı tried to understand but not perfect.
```

Comment:
It's very similar to the exercise above. We need to iterate through each key of the object and check if a value is object or not. If a value of the key is an object, we need to run the function recursively there but also increase the depth to keep track of it. We run function recursively until there are no more objects as values.